core/store/orm/orm.go

```diff
@@ -72,14 +78,46 @@ func NewORM(uri string, timeout time.Duration) (*ORM, error) {
 	}
 
 	orm := &ORM{
-		DB:              db,
-		lockingStrategy: lockingStrategy,
-		dialectName:     dialect,
+		DB:                 db,
+		lockingStrategy:    lockingStrategy,
+		acquireLockTimeout: timeout,
+		dialectName:        dialect,
+	}
+
+	if dialect == DialectPostgres {
+		db.Callback().Create().Before("gorm:begin_transaction").Register("chainlink:before_create", orm.acquireAdvisoryLock)
+		db.Callback().Create().After("gorm:commit_or_rollback_transaction").Register("chainlink:after_create", orm.releaseAdvisoryLock)
+
+		db.Callback().Delete().Before("gorm:begin_transaction").Register("chainlink:before_delete", orm.acquireAdvisoryLock)
+		db.Callback().Delete().After("gorm:commit_or_rollback_transaction").Register("chainlink:after_delete", orm.releaseAdvisoryLock)
+
+		db.Callback().Update().Before("gorm:begin_transaction").Register("chainlink:before_update", orm.acquireAdvisoryLock)
+		db.Callback().Update().After("gorm:commit_or_rollback_transaction").Register("chainlink:after_update", orm.releaseAdvisoryLock)
+
+		db.Callback().Query().Before("gorm:query").Register("chainlink:before_query", orm.acquireAdvisoryLock)
+		db.Callback().Query().After("gorm:after_query").Register("chainlink:after_query", orm.releaseAdvisoryLock)
+
+		db.Callback().RowQuery().Before("gorm:row_query").Register("chainlink:before_row_query", orm.acquireAdvisoryLock)
+		db.Callback().RowQuery().After("gorm:row_query").Register("chainlink:after_row_query", orm.releaseAdvisoryLock)
 	}
 
 	return orm, nil
 }
 
+func (orm *ORM) acquireAdvisoryLock(scope *gorm.Scope) {
+	err := orm.lockingStrategy.Lock(orm.acquireLockTimeout)
+	if err != nil {
+		scope.Err(errors.Wrap(ErrNoAdvisoryLock, err.Error()))
```

Comment:
With this current implementation, the chainlink node will log an error and continue running. Is this our intent @jleeh or do we ultimately want to stop the CL node entirely? I see 2 CL nodes continuing to run with one spamming logs.

Perhaps you already have an answer to this @spooktheducks, feel free to capture it here.

Reply:
Locally, I found that the node actually died in this case.  It's also the only in-band way that GORM gives the user to issue errors from callbacks.  If I'm wrong about the behavior I thought I saw, we can just wrap GORM's DB object directly.

Reply:
Locally, I found that the node actually died in this case. It's also the only in-band way that GORM gives the user to issue errors from callbacks. If I'm wrong about the behavior I thought I saw, we can just wrap GORM's DB object directly.